src/librbd/api/Trash.cc

```diff
@@ -171,6 +170,28 @@ int Trash<I>::move(librados::IoCtx &io_ctx, rbd_trash_image_source_t source,
     return r;
   }
 
+  //migration already issues a request to check for watchers
+  if(source != RBD_TRASH_IMAGE_SOURCE_MIGRATION) {
+    std::list<obj_watch_t> t_watchers;
+    int flags = librbd::image::LIST_WATCHERS_FILTER_OUT_MY_INSTANCE |
+                librbd::image::LIST_WATCHERS_FILTER_OUT_MIRROR_INSTANCES;
+    C_SaferCond t_on_list_watchers;
+    auto list_req = librbd::image::ListWatchersRequest<I>::create(
+        *ictx, flags, &t_watchers, &t_on_list_watchers);
+    list_req->send();
+    r = t_on_list_watchers.wait();
+    if (r < 0) {
+      lderr(cct) << "failed listing watchers:" << cpp_strerror(r) << dendl;
+      ictx->state->close();
+      return r;
+    }
+    if (!t_watchers.empty()) {
+      lderr(cct) << "image has watchers - not moving" << dendl;
+      ictx->state->close();
+      return -EBUSY;
+    }
+  }
+
```

Comment:
Again, not exactly related, but shouldn't we close the image at this stage?

Reply:
Image gets closed below - https://github.com/ceph/ceph/blob/master/src/librbd/api/Trash.cc#L188